pkg/tbtc/tbtc.go

```diff
@@ -146,16 +146,69 @@ func Initialize(
 				return
 			}
 
+			confirmationBlock := event.BlockNumber + dkgStartedConfirmationBlocks
+
 			logger.Infof(
-				"DKG started with seed [0x%x] at block [%v]",
+				"observed DKG started event with seed [0x%x] and "+
+					"starting block [%v]; waiting for block [%v] to confirm",
 				event.Seed,
 				event.BlockNumber,
+				confirmationBlock,
 			)
 
-			node.joinDKGIfEligible(
-				event.Seed,
-				event.BlockNumber,
-			)
+			err := node.waitForBlockHeight(ctx, confirmationBlock)
+			if err != nil {
+				logger.Errorf("failed to confirm DKG started event: [%v]", err)
+				return
+			}
+
+			dkgState, err := chain.GetDKGState()
+			if err != nil {
+				logger.Errorf("failed to check DKG state: [%v]", err)
+				return
+			}
+
+			if dkgState == AwaitingResult {
+				// Fetch all past DKG started events starting from one
+				// confirmation period before the original event's block.
+				pastEvents, err := chain.PastDKGStartedEvents(
+					&DKGStartedEventFilter{
+						StartBlock: event.BlockNumber - dkgStartedConfirmationBlocks,
+					},
+				)
+				if err != nil {
+					logger.Errorf("failed to get past DKG started events: [%v]", err)
+					return
+				}
+
+				lastEvent := pastEvents[len(pastEvents)-1]
+
+				logger.Infof(
+					"DKG started with seed [0x%x] at block [%v]",
+					lastEvent.Seed,
+					lastEvent.BlockNumber,
+				)
+
+				// The off-chain protocol should be started as close as possible
+				// to the current block or even further. Starting the off-chain
+				// protocol with a past block will likely cause a failure of the
+				// first attempt as the start block is used to synchronize
+				// the announcements and the state machine. Here we ensure
+				// a proper start point by delaying the execution by the
+				// confirmation period length.
+				node.joinDKGIfEligible(
+					lastEvent.Seed,
+					lastEvent.BlockNumber,
+					dkgStartedConfirmationBlocks,
+				)
+			} else {
+				logger.Infof(
+					"DKG started event with seed [0x%x] and starting "+
+						"block [%v] was not confirmed",
+					event.Seed,
+					event.BlockNumber,
+				)
+			}
```

Comment:
Not blocking this PR but the entire content of `OnDKGStarted` is getting quite complicated. We could consider extracting it to a separate function in `tbtc.go` to unit-test it. Maybe a separate PR?

Reply:
Sure! I can do it during the work on https://github.com/keep-network/keep-core/issues/3462